reference/reference.go

```diff
@@ -56,6 +65,11 @@ var (
 type Reference interface {
 	// String returns the full reference
 	String() string
+
+	// internal requires that all instances of Reference must be implemented
+	// by this package. This allows us to ensure that instances can be treated
+	// as an algebraic data type.
+	internal()
```

Comment:
I see what you're going for here but I don't think we should make this interface change yet, because it will make it impossible to vendor an updated distribution repository in the Docker Engine until all the reference changes are upstreamed. We will need to vendor an updated distribution repository soon to get the updated manifest code. So I would prefer to wait until the Engine is ported over to a unified `reference` package before adding a `internal()` method to this interface.